src/d3d9/d3d9_shader_validator.h

```diff
@@ -37,32 +71,203 @@ namespace dxvk {
+        return E_FAIL;
+      }
+
+      if (m_state == D3D9ShaderValidatorState::ValidatingHeader)
+        return ValidateHeader(pFile, Line, pdwInst, cdw);
+
+      DxsoCodeIter pdwInstIter{ reinterpret_cast<const uint32_t*>(pdwInst) };
+      bool isEndToken = !m_ctx->decodeInstruction(pdwInstIter);
+      const DxsoInstructionContext instContext = m_ctx->getInstructionContext();
+
+      if (isEndToken)
+        return ValidateEndToken(pFile, Line, pdwInst, cdw);
+
+      // TODO: DxsoDecodeContext::decodeInstructionLength() does not currently appear
+      // to return the correct token length in many cases, and as such dwordLength
+      // will not be equal to cdw in many situations that are expected to pass validation
+      //
+      /*Logger::debug(str::format("IDirect3DShaderValidator9::Instruction: opcode ", instContext.instruction.opcode));
+      // + 1 to account for the opcode...
+      uint32_t dwordLength = instContext.instruction.tokenLength + 1;
+      Logger::debug(str::format("IDirect3DShaderValidator9::Instruction: cdw ", cdw));
+      Logger::debug(str::format("IDirect3DShaderValidator9::Instruction: dwordLength ", dwordLength));
+      if (unlikely(cdw != dwordLength)) {
+        return ErrorCallback(pFile, Line, 0x2,
+          D3D9ShaderValidatorMessage::BadInstructionLength,
+          str::format("Instruction length specified for instruction (", cdw, ") does not match the token count encountered (", dwordLength, "). Aborting validation."));
+      }*/
+
+      // a maximum of 10 inputs are supported with PS 3.0 (validation required by The Void)
+      if (m_isPixelShader && m_majorVersion == 3) {
+        switch (instContext.instruction.opcode) {
+          case DxsoOpcode::Comment:
+          case DxsoOpcode::Def:
+          case DxsoOpcode::DefB:
+          case DxsoOpcode::DefI:
+            break;
+
+          default:
+            for (uint32_t instNum = 1; pdwInst[instNum] & 0x80000000; instNum++) {
+              DWORD regType  = ((pdwInst[instNum] & D3DSP_REGTYPE_MASK)  >> D3DSP_REGTYPE_SHIFT)
+                             | ((pdwInst[instNum] & D3DSP_REGTYPE_MASK2) >> D3DSP_REGTYPE_SHIFT2);
+              DWORD regIndex = pdwInst[instNum] & D3DSP_REGNUM_MASK;
+
+              if (unlikely(regType == D3DSPR_INPUT && regIndex >= 10)) {
+                return ErrorCallback(pFile, Line, 0x2,
+                  instContext.instruction.opcode == DxsoOpcode::Dcl ?
+                    D3D9ShaderValidatorMessage::BadInputRegisterDeclaration :
+                    D3D9ShaderValidatorMessage::BadInputRegister,
+                  "IDirect3DShaderValidator9::Instruction: Invalid number of PS input registers specified. Aborting validation.");
+              }
+            }
+        }
+      }
 
       return D3D_OK;
     }
 
 
     HRESULT STDMETHODCALLTYPE End() {
-      Logger::debug("D3D9ShaderValidator::End: Stub");
+      if (unlikely(m_state == D3D9ShaderValidatorState::Error)) {
```

Comment:
A good question for @misyltoad . I'm afraid I have no idea how to go about testing this. Games do seem to recreate the validator instance for each shader from what I've noticed, but I have no idea if this is intended by design.

Reply:
P.S.: According to [these tests](https://bugs.winehq.org/attachment.cgi?id=69141) it does indeed seem that once a validation error is hit, the validator instances must be released and recreated. Great API, I know.

Reply:
So it can be reused if the shader is fine but not if there's issues with it?

Reply:
> So it can be reused if the shader is fine but not if there's issues with it?

Appears that way, yes. I think I'll end up writing a very targeted test on this behavior by "borrowing" some valid/invalid definitions from the Wine tests.